pennylane/operation.py

```diff
@@ -232,6 +232,91 @@ def expand_matrix(base_matrix, wires, wire_order=None):
     return mat
 
 
+def _local_sparse_swap_mat(i, n, format="csr"):
+    """Helper function which generates the sparse matrix of SWAP
+    for qubits: i <--> i+1 with final shape (2**n, 2**n)."""
+    assert i < n - 1
+    swap_mat = csr_matrix([[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 0, 1]])
+
+    j = i + 1  # i is the index of the qubit, j is the number of qubits prior to and include qubit i
+    return kron(
+        kron(eye(2 ** (j - 1)), swap_mat), eye(2 ** (n - (j + 1))), format=format
+    )  # (j - 1) + 2 + (n - (j+1)) = n
+
+
+def _sparse_swap_mat(i, j, n, format="csr"):
+    """Helper function which generates the sparse matrix of SWAP
+    for qubits: i <--> j with final shape (2**n, 2**n)."""
+    assert i < n and j < n
+    if i == j:
+        return eye(2**n, format=format)
+
+    (small_i, big_j) = (i, j) if i < j else (j, i)
+    store_swaps = [
+        _local_sparse_swap_mat(index, n, format=format) for index in range(small_i, big_j)
+    ]
+
+    res = eye(2**n, format=format)
+    for mat in store_swaps:  # swap i --> j
+        res @= mat
+
+    for mat in store_swaps[-2::-1]:  # bring j --> old_i
+        res @= mat
+
+    return res
+
+
+def sparse_expand_matrix(base_matrix, wires, wire_order=None, format="csr"):
+    """Re-express a sparse base matrix acting on a subspace defined by a set of wire labels
+    according to a global wire order.
+
+    Args:
+        base_matrix (scipy.sparse.spmatrix): base matrix to expand
+        wires (Iterable): wires determining the subspace that base matrix acts on; a base matrix of
+            dimension :math:`2^n` acts on a subspace of :math:`n` wires
+        wire_order (Iterable): global wire order, which has to contain all wire labels in ``wires``, but can also
+            contain additional labels
+        format (str): string representing the preferred scipy sparse matrix format to cast the expanded
+            matrix too
+
+    Returns:
+        tensor_like: expanded matrix
+    """
+
+    if (wire_order is None) or (wire_order == wires):
+        return base_matrix
+
+    interface = qml.math.get_interface(base_matrix)  # pylint: disable=protected-access
+    if interface != "scipy" and not issparse(base_matrix):
+        raise ValueError("base_matrix must be a scipy sparse matrix")
+
+    n_wires = len(wires)
+    n_total_wires = len(wire_order)
+
+    expanded_wires = copy.copy(wires)
+    for wire in wire_order:
+        if wire not in wires:
+            expanded_wires.append(wire)
+
+    num_missing_wires = n_total_wires - n_wires
+    expanded_matrix = kron(
+        base_matrix, eye(2**num_missing_wires, format=format), format=format
+    )  # added missing wires at the end
+
+    U = eye(2**n_total_wires)
+    while not expanded_wires == wire_order:
+        for i in range(n_total_wires):
```

Comment:
Thinking about this a bit more, are you sure this outer `while` loop is necessary? It looks like after the `i`th iteration of the inner loop, all of `expanded_wires[0..i]` are now in the correct position.

Of course, I'd concede this if you find some counterexample.

Reply:
I know I came across a case in the tests where not including the while loop resulted in an incorrect permutation. Let me double check and see if I can't find it again

Reply:
Consider the case `expanded_wires = [1, 2, 3, 0]` and `wire_order = [0, 1, 2, 3]`. The first index check will swap 1 <--> 2. The second index check will pass. The third index check will swap 3 <--> 0. Resulting in `expanded_wires = [2, 1, 0, 3]`

Reply:
Right, how about instead of
```python
j = wire_order.index(expanded_wires[i])
```
we have
```python
j = expanded_wires.index(wire_order[i])
```
and remove the `break` statement. I think this way we maintain the loop invariant that `expanded_wires[0:i + 1] == wire_order[0:i + 1]` after iteration `i`.

Reply:
Nice! I think this should work in all cases!

Reply:
Added 👍🏼